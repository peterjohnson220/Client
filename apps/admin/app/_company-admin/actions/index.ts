export * from './navigation.actions';
export * from './password-management-settings.action';
export * from './user-role-view.action';
export * from './user-role-users-tab.action';
export * from './user-role-functions-tab.action';
export * from './data-access-tab.action';
export * from './user.actions';
