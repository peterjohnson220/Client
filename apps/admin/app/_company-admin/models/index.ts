export * from './add-company-role-form.model';
