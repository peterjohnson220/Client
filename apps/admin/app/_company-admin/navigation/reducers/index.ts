import { createFeatureSelector, createSelector } from '@ngrx/store';

// Import root app reducer
import * as fromRoot from 'libs/state/state';

// Import feature reducers
import * as fromNavigationReducer from './navigation.reducer';

// Feature area state
export interface CompanyAdminNavigationState {
  navigation: fromNavigationReducer.State;
}

// Extend root state with feature area state
export interface State extends fromRoot.State {
  companyAdminNavigation: CompanyAdminNavigationState;
}

// Feature area reducers
export const reducers = {
  navigation: fromNavigationReducer.reducer,
};

// Select Feature Area
export const selectCompanyAdminNavigationState =
  createFeatureSelector<CompanyAdminNavigationState>('companyAdminNavigation');

// View Selectors
export const selectNavigationState =
  createSelector(selectCompanyAdminNavigationState, (state: CompanyAdminNavigationState) => state.navigation);

// Navigation Links
export const getLinks = createSelector(
  selectNavigationState, fromNavigationReducer.getLinks
);

export const getNavigationLinksLoading = createSelector(
  selectNavigationState, fromNavigationReducer.getLoading
);

export const getNavigationLinksLoadingError = createSelector(
  selectNavigationState, fromNavigationReducer.getLoadingError
);

