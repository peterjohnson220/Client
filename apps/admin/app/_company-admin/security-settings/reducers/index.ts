import { createFeatureSelector, createSelector } from '@ngrx/store';

// Import root app reducer
import * as fromRoot from 'libs/state/state';

// Import feature reducers
import * as fromPasswordSettingsReducer from './security-setting.reducer';

// Feature area state
export interface CompanyAdminPasswordSettingsState {
  passwordSettings: fromPasswordSettingsReducer.State;
}

// Extend root state with feature area state
export interface State extends fromRoot.State {
  companyAdminPasswordSettings: CompanyAdminPasswordSettingsState;
}

// Feature area reducers
export const reducers = {
  passwordSettings: fromPasswordSettingsReducer.reducer,
};

// Select Feature Area
export const selectCompanyAdminPasswordSettingsState =
  createFeatureSelector<CompanyAdminPasswordSettingsState>('companyAdminPasswordSettings');

// View Selectors

export const selectPasswordSettingsState =
  createSelector(selectCompanyAdminPasswordSettingsState, (state: CompanyAdminPasswordSettingsState) => state.passwordSettings);

// Password Settings
export const getPasswordSettings = createSelector(
  selectPasswordSettingsState, fromPasswordSettingsReducer.getPasswordSettings
);

export const getPasswordSettingsLoading = createSelector(
  selectPasswordSettingsState, fromPasswordSettingsReducer.getLoading
);

export const getPasswordSettingsLoadingError = createSelector(
  selectPasswordSettingsState, fromPasswordSettingsReducer.getLoadingError
);

export const getPasswordSettingsModalOpen = createSelector(
  selectPasswordSettingsState, fromPasswordSettingsReducer.getSaveModalOpen
);

export const getPasswordSettingsSaveRequest = createSelector(
  selectPasswordSettingsState, fromPasswordSettingsReducer.getSaveRequest
);

export const getPasswordSettingsSaving = createSelector(
  selectPasswordSettingsState, fromPasswordSettingsReducer.getSaving
);

export const getPasswordSettingsSavingError = createSelector(
  selectPasswordSettingsState, fromPasswordSettingsReducer.getSavingError
);
