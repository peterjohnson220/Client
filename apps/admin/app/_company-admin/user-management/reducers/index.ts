import { createFeatureSelector, createSelector } from '@ngrx/store';

// Import root app reducer
import * as fromRoot from 'libs/state/state';

// Import feature reducers
import * as fromUserManagementReducer from './user-management.reducer';

// Feature area state
export interface CompanyAdminUserManagementState {
  userForm: fromUserManagementReducer.IUserState;
}

// Extend root state with feature area state
export interface State extends fromRoot.State {
  companyAdminUserManagement: CompanyAdminUserManagementState;
}

// Feature area reducers
export const reducers = {
  userForm: fromUserManagementReducer.reducer
};

// Select Feature Area
export const selectCompanyAdminUserManagementState =
  createFeatureSelector<CompanyAdminUserManagementState>('companyAdminUserManagement');

// View Selectors
export const selectUserFormState =
  createSelector(selectCompanyAdminUserManagementState, (state: CompanyAdminUserManagementState) => state.userForm);


// User Form
export const getUserState = createSelector(
  selectUserFormState, fromUserManagementReducer.getUserState
);

export const getUserStateRoles = createSelector(
  selectUserFormState, fromUserManagementReducer.getRoles
);

export const getUserStateUser = createSelector(
  selectUserFormState, fromUserManagementReducer.getUser
);

export const getUserStateLoading = createSelector(
  selectUserFormState, fromUserManagementReducer.getUserLoading
);

export const getUserStateLoaded = createSelector(
  selectUserFormState, fromUserManagementReducer.getUserLoaded
);

export const getUserStateApiError = createSelector(
  selectUserFormState, fromUserManagementReducer.getUserApiError
);
