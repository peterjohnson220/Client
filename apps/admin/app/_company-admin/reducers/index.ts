import { createFeatureSelector, createSelector } from '@ngrx/store';

// Import root app reducer
import * as fromRoot from 'libs/state/state';

// Import feature reducers
import * as fromUserRoleViewReducer from './user-role-view.reducer';
import * as fromUserRoleUserTabReducer from './user-role-users-tab.reducer';
import * as fromUserRoleFunctionTabReducer from './user-role-functions-tab.reducer';
import * as fromNavigationReducer from './navigation.reducer';
import * as fromDataAccessTabReducer from './role-data-access-tab.reducer';

// Feature area state
export interface UserRoleViewStateMain {
  userRoleView: fromUserRoleViewReducer.IUserRoleState;
  userRoleUserTab: fromUserRoleUserTabReducer.State;
  userRoleFunctionTab: fromUserRoleFunctionTabReducer.State;
  navigation: fromNavigationReducer.State;
  dataAccessTab: fromDataAccessTabReducer.State;
}

// Extend root state with feature area state
export interface State extends fromRoot.State {
  userRoleAdminMain: UserRoleViewStateMain;
}

// Feature area reducers
export const reducers = {
  userRoleView: fromUserRoleViewReducer.reducer,
  userRoleUserTab: fromUserRoleUserTabReducer.reducer,
  userRoleFunctionTab: fromUserRoleFunctionTabReducer.reducer,
  navigation: fromNavigationReducer.reducer,
  dataAccessTab: fromDataAccessTabReducer.reducer
};

// Select Feature Area
export const selectUserRoleAdminMainState =
  createFeatureSelector<UserRoleViewStateMain>('userRoleAdminMain');

// User Role View Selectors
export const selectUserRoleState =
  createSelector(selectUserRoleAdminMainState, (state: UserRoleViewStateMain) => state.userRoleView);

export const selectUserRoleFunctionTabState =
  createSelector(selectUserRoleAdminMainState, (state: UserRoleViewStateMain) => state.userRoleFunctionTab);

export const selectRoleDataAccessTabState =
  createSelector(selectUserRoleAdminMainState, (state: UserRoleViewStateMain) => state.dataAccessTab);

export const selectUserRoleUserTabState =
  createSelector(selectUserRoleAdminMainState, (state: UserRoleViewStateMain) => state.userRoleUserTab);

export const selectNavigationState =
  createSelector(selectUserRoleAdminMainState, (state: UserRoleViewStateMain) => state.navigation);

export const getUserRoleViewState = createSelector(
  selectUserRoleState, fromUserRoleViewReducer.getUserRoleViewState
);

export const getUserRoleCurrentTabState = createSelector(
  selectUserRoleState, fromUserRoleViewReducer.getUserRoleCurrentTab
);

export const getCurrentUserRole = createSelector(
  selectUserRoleState, fromUserRoleViewReducer.getCurrentUserRole
);

export const getCompanyRoles = createSelector(
  selectUserRoleState, fromUserRoleViewReducer.getCompanyRoles
);

export const getAddCompanyRoleModalIsOpen = createSelector(
  selectUserRoleState, fromUserRoleViewReducer.getAddCompanyRoleModalIsOpen
);

export const getAddCompanyRoleForm = createSelector(
  selectUserRoleState, fromUserRoleViewReducer.getAddCompanyRoleForm
);

export const getAddCompanyRoleError = createSelector(
  selectUserRoleState, fromUserRoleViewReducer.getAddCompanyRoleError
);

export const getRoleApiResponse = createSelector(
  selectUserRoleState, fromUserRoleViewReducer.getRoleApiResponse
);

export const getSaveButtonDisabled = createSelector(
  selectUserRoleState, fromUserRoleViewReducer.getSaveButtonDisabled
);

// Functions tab
export const getFunctionTabPermissions = createSelector(
  selectUserRoleFunctionTabState, fromUserRoleFunctionTabReducer.getFunctionTabPermissions
);

export const getFunctionTabPendingChanges = createSelector(
  selectUserRoleFunctionTabState, fromUserRoleFunctionTabReducer.getFunctionTabPendingChanges
);

export const getCurrentCheckedPermissionIds = createSelector(
  selectUserRoleFunctionTabState, fromUserRoleFunctionTabReducer.getCheckedPermissions
);

export const getCheckboxesDisabled = createSelector(
  selectUserRoleFunctionTabState, fromUserRoleFunctionTabReducer.getCheckboxesDisabled
);

// Users tab
export const getUsersAndRoles = createSelector(
  selectUserRoleUserTabState, fromUserRoleUserTabReducer.getUsersAndRoles
);

export const getUsersAndRolesError = createSelector(
  selectUserRoleUserTabState, fromUserRoleUserTabReducer.getUsersAndRolesError
);

export const getUsersInActiveRole = createSelector(
  selectUserRoleUserTabState, fromUserRoleUserTabReducer.getUsersInActiveRole
);

export const getUsersNotInActiveRole = createSelector(
  selectUserRoleUserTabState, fromUserRoleUserTabReducer.getUsersNotInActiveRole
);

export const getUsersTabHasPendingChanges = createSelector(
  selectUserRoleUserTabState, fromUserRoleUserTabReducer.getUsersTabHasPendingChanges
);

export const getUserIdsToSave = createSelector(
  selectUserRoleUserTabState, fromUserRoleUserTabReducer.getUserIdsToSave
);

// Navigation Links
export const getLinks = createSelector(
  selectNavigationState, fromNavigationReducer.getLinks
);

export const getNavigationLinksLoading = createSelector(
  selectNavigationState, fromNavigationReducer.getLoading
);

export const getNavigationLinksLoadingError = createSelector(
  selectNavigationState, fromNavigationReducer.getLoadingError
);

// Data Access Tab

export const getDataTypes = createSelector(
  selectRoleDataAccessTabState, fromDataAccessTabReducer.getDataTypes
);

export const getRoleDataRestrictions = createSelector(
  selectRoleDataAccessTabState, fromDataAccessTabReducer.getRoleDataRestrictions
);

export  const getDataAccessTabPendingChanges = createSelector(
  selectRoleDataAccessTabState, fromDataAccessTabReducer.getDataAccessTabHasPendingChanges
);
