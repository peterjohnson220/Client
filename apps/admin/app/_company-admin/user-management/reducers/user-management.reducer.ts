import * as fromUserManagementActions from '../actions/user-management.actions';
import { UserAssignedRole } from 'libs/models';
import { UserManagementDto } from 'libs/models/payfactors-api/user';

export interface IUserState {
    loaded: boolean;
    loading: boolean;
    user: UserManagementDto;
    roles: UserAssignedRole[];
    apiError: string;

}

const initialState: IUserState = {
    loaded: false,
    loading: false,
    user: null,
    roles: null,
    apiError: ''
};

export function reducer(state = initialState, action: fromUserManagementActions.UserManagementActions): IUserState {
    const noApiResponseState = {
        ...state,
        apiSaveResponse: undefined
    };
    switch (action.type) {
        case fromUserManagementActions.LOAD_USER:
            return {
                ...noApiResponseState,
                loading: true
            };
        case fromUserManagementActions.LOAD_USER_SUCCESS:
            return {
                ...noApiResponseState,
                loaded: true,
                loading: false,
                user: action.payload,
            };
        case fromUserManagementActions.RESET_USER:
            return {
                ...noApiResponseState,
                user: null,
            };
        case fromUserManagementActions.LOAD_ROLES_SUCCESS:
            return {
                ...noApiResponseState,
                roles: action.payload,
            };
        case fromUserManagementActions.SAVE_USER:
            return {
                ...noApiResponseState,
                loading: true
            };
        case fromUserManagementActions.SAVE_USER_SUCCESS:
            return {
                ...noApiResponseState,
                loading: false
            };
        case fromUserManagementActions.HANDLE_API_ERROR:
            return {
                ...noApiResponseState,
                loading: false,
                apiError: action.payload
            };
        default:
            return state;
    }
}

export const getUserState = (state: IUserState) => state;
export const getUser = (state: IUserState) => state.user;
export const getRoles = (state: IUserState) => state.roles;
export const getUserLoading = (state: IUserState) => state.loading;
export const getUserLoaded = (state: IUserState) => state.loaded;
export const getUserApiError = (state: IUserState) => state.apiError;
