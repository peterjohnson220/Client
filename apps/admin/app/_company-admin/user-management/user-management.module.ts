import { NgModule} from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';

import { StoreModule } from '@ngrx/store';
import { EffectsModule } from '@ngrx/effects';
import { GridModule } from '@progress/kendo-angular-grid';
import { LayoutModule } from '@progress/kendo-angular-layout';
import { FontAwesomeModule } from '@fortawesome/angular-fontawesome';

import { PfCommonUIModule } from 'libs/ui/common';
import { PfFormsModule } from 'libs/forms';

import { UserEffects } from './effects/user-management.effects';
import { reducers } from './reducers';

import { UserFormComponent } from './components/user-form';
import { UserManagementPageComponent } from './user-management.page/user-management.page';


@NgModule({
  imports: [
    // Angular
    CommonModule, FormsModule, ReactiveFormsModule,

    // 3rd Party
    StoreModule.forFeature('companyAdminUserManagement', reducers),
    EffectsModule.forFeature([
      UserEffects,
    ]),
    GridModule,
    LayoutModule,
    FontAwesomeModule,

    // Payfactors
    PfCommonUIModule,
    PfFormsModule,
  ],
  declarations: [
    // Feature
    UserManagementPageComponent,

    // Components
    UserFormComponent
  ],
  exports: [
    UserManagementPageComponent,
  ]
})
export class UserManagementModule {
  constructor() { }
}
