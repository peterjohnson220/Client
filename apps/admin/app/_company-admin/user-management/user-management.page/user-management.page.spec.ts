import { NO_ERRORS_SCHEMA } from '@angular/core';
import { ComponentFixture, TestBed, async, getTestBed } from '@angular/core/testing';
import { Router } from '@angular/router';
import { Location } from '@angular/common';
import { RouterTestingModule } from '@angular/router/testing';

import { combineReducers, Store, StoreModule } from '@ngrx/store';

import { Observable } from 'rxjs';
import 'rxjs/add/observable/of';

import * as fromRootState from 'libs/state/state';
import * as fromUserReducer from '../reducers';
import * as fromUserActions from '../actions/user-management.actions';

import { UserManagementPageComponent } from './user-management.page';
import { generateMockUserManagementDto } from 'libs/models/payfactors-api/user';
import { generateMockUserContext } from 'libs/models';
import { Permissions } from 'libs/constants';

describe('Admin - Company Admin - User Page', () => {
  let instance: UserManagementPageComponent;
  let fixture: ComponentFixture<UserManagementPageComponent>;
  let store: Store<fromUserReducer.State>;
  let router: Router;
  let location: Location;

  const mockCompanyId = 13;
  const mockUserId = 25;

  beforeEach(async(() => {

    TestBed.configureTestingModule({
      imports: [
        StoreModule.forRoot({
          ...fromRootState.reducers,
          companyAdminUserManagement: combineReducers(fromUserReducer.reducers),
        }),
        RouterTestingModule.withRoutes(
          [
            { path: ':companyId/users/add', component: UserManagementPageComponent },
            { path: ':companyId/users/:userId', component: UserManagementPageComponent },
          ]
        ),
      ],
      declarations: [
        UserManagementPageComponent,
      ],
      schemas: [NO_ERRORS_SCHEMA]
    }).compileComponents();

    const injector = getTestBed();
    location = injector.get(Location);
    router = injector.get(Router);

    store = TestBed.get(Store);
    spyOn(store, 'dispatch');

    fixture = TestBed.createComponent(UserManagementPageComponent);

    const mockUserContext = generateMockUserContext();
    mockUserContext.Permissions.push(Permissions.ADD_USER);

    instance = fixture.componentInstance;
    instance.userContext$ = Observable.of(mockUserContext);

  }));

  it('Should show the add user page', () => {
    instance.userId = null;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('Should show the edit user page', () => {
    instance.userId = mockUserId;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('Should dispatch an LoadRoles action on initial load', () => {
    const expectedAction = new fromUserActions.LoadRoles(mockCompanyId);
    router.navigate([`${this.mockCompanyId}/users/${this.mockUserId}`]).then(() => {
      expect(store.dispatch).toHaveBeenCalledWith(expectedAction);
    });
  });

  it('Should dispatch an LoadUser action for edits', () => {
    const expectedAction = new fromUserActions.LoadUser(mockUserId);
    router.navigate([`${this.mockCompanyId}/users/${this.mockUserId}`]).then(() => {
      expect(store.dispatch).toHaveBeenCalledWith(expectedAction);
    });
  });

  it('Should dispatch an ResetUser action for adding new user', () => {
    const expectedAction = new fromUserActions.ResetUser();
    router.navigate([`${this.mockCompanyId}/users/add`]).then(() => {
      expect(store.dispatch).toHaveBeenCalledWith(expectedAction);
    });
  });

  it('Should dispatch an SaveUser action on Save', () => {
    const mockUser = generateMockUserManagementDto();
    const expectedAction = new fromUserActions.SaveUser(mockUser);

    instance.save(mockUser);

    expect(store.dispatch).toHaveBeenCalledWith(expectedAction);
  });

  it('Should navigate to the user list page when the cancel button is clicked', () => {
    spyOn(location, 'back');
    instance.handleCancel();
    expect(location.back).toHaveBeenCalled();
  });

});
