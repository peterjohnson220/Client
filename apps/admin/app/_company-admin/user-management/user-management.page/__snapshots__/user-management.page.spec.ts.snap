// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admin - Company Admin - User Page Should show the add user page 1`] = `
<pf-user-management-page
  accessLevel={[Function String]}
  addUserFlag={[Function Boolean]}
  apiError$={[Function Store]}
  companyId="undefined"
  loaded$={[Function Store]}
  loading$={[Function Store]}
  location={[Function SpyLocation]}
  permissions={[Function Array]}
  route={[Function ActivatedRoute]}
  store={[Function Store]}
  user$={[Function Store]}
  userContext$={[Function Observable]}
  userId="undefined"
  userRoles$={[Function Store]}
>
  <pf-async-container>
    <pf-page-with-title-bar>
      
       Add User 
      
      <button
        class="btn btn-primary"
        id="user-page-save"
      >
        Save
      </button>
      <button
        class="btn btn-secondary"
        id="user-page-cancel"
      >
        Cancel
      </button>
      
      
      <div
        class="row no-gutters"
      >
        <div
          class="col-lg-6 px-3"
        >
          <pf-user-form />
        </div>
        <div
          class="col-lg-6 px-3 mt-4"
        >
          <div
            class="card"
          >
            <div
              class="card-body"
            >
              <h3
                class="card-title text-center"
              >
                 We've moved! 
              </h3>
              <p
                class="card-text"
              >
                 Tile access and restrictions are now managed with our new User Roles functionality. To access User Roles, click on the dropdown next to your name on the top right of the screen and select Company Administration. In the User Management section, you will see a User Roles link if you have access. Use User Roles to create permission roles and assign your users to those roles. 
              </p>
            </div>
          </div>
        </div>
      </div>
    </pf-page-with-title-bar>
  </pf-async-container>
</pf-user-management-page>
`;

exports[`Admin - Company Admin - User Page Should show the edit user page 1`] = `
<pf-user-management-page
  accessLevel={[Function String]}
  addUserFlag={[Function Boolean]}
  apiError$={[Function Store]}
  companyId="undefined"
  loaded$={[Function Store]}
  loading$={[Function Store]}
  location={[Function SpyLocation]}
  permissions={[Function Array]}
  route={[Function ActivatedRoute]}
  store={[Function Store]}
  user$={[Function Store]}
  userContext$={[Function Observable]}
  userId="undefined"
  userRoles$={[Function Store]}
>
  <pf-async-container>
    <pf-page-with-title-bar>
      
       Add User 
      
      <button
        class="btn btn-primary"
        id="user-page-save"
      >
        Save
      </button>
      <button
        class="btn btn-secondary"
        id="user-page-cancel"
      >
        Cancel
      </button>
      
      
      <div
        class="row no-gutters"
      >
        <div
          class="col-lg-6 px-3"
        >
          <pf-user-form />
        </div>
        <div
          class="col-lg-6 px-3 mt-4"
        >
          <div
            class="card"
          >
            <div
              class="card-body"
            >
              <h3
                class="card-title text-center"
              >
                 We've moved! 
              </h3>
              <p
                class="card-text"
              >
                 Tile access and restrictions are now managed with our new User Roles functionality. To access User Roles, click on the dropdown next to your name on the top right of the screen and select Company Administration. In the User Management section, you will see a User Roles link if you have access. Use User Roles to create permission roles and assign your users to those roles. 
              </p>
            </div>
          </div>
        </div>
      </div>
    </pf-page-with-title-bar>
  </pf-async-container>
</pf-user-management-page>
`;
